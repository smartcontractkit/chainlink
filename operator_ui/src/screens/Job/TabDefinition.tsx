import React from 'react'
import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter'
import { prism } from 'react-syntax-highlighter/dist/esm/styles/prism'

import Card from '@material-ui/core/Card'
import CardContent from '@material-ui/core/CardContent'
import CardHeader from '@material-ui/core/CardHeader'
import Typography from '@material-ui/core/Typography'

import { CopyButton } from 'src/components/Copy/CopyButton'
import { generateJobDefinition } from './generateJobDefinition'

interface Props {
  job: JobPayload_Fields
}

export const TabDefinition = ({ job }: Props) => {
  const { definition, envDefinition } = generateJobDefinition(job)

  return (
    <Card>
      <CardHeader
        title="Definition"
        action={<CopyButton title="Copy" data={definition} />}
      />
      <CardContent>
        <Typography style={{ margin: 0 }} variant="body1" component="pre">
          <SyntaxHighlighter
            language="toml"
            style={prism}
            data-testid="definition"
          >
            {definition}
          </SyntaxHighlighter>
        </Typography>

        {envDefinition.trim() && (
          <>
            <Typography variant="h5" color="secondary">
              Job attributes set by environment variables
            </Typography>

            <Typography style={{ margin: 0 }} variant="body1" component="pre">
              <SyntaxHighlighter language="toml" style={prism}>
                {envDefinition}
              </SyntaxHighlighter>
            </Typography>
          </>
        )}
      </CardContent>
    </Card>
  )
}
