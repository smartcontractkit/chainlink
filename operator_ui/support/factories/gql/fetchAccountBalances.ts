// buildETHKey builds a eth key for the FetchETHKeys query.
export function buildETHKey(
  overrides?: Partial<AccountBalancesPayload_ResultsFields>,
): AccountBalancesPayload_ResultsFields {
  return {
    __typename: 'EthKey',
    address: '0x0000000000000000000000000000000000000001',
    chain: {
      __typename: 'Chain',
      id: '42',
    },
    ethBalance: '0.100000000000000000',
    isFunding: false,
    linkBalance: '1000000000000000000',
    ...overrides,
  }
}

// buildETHKeys builds a list of eth keys.
export function buildETHKeys(): ReadonlyArray<AccountBalancesPayload_ResultsFields> {
  return [
    buildETHKey(),
    buildETHKey({
      address: '0x0000000000000000000000000000000000000002',
    }),
  ]
}
