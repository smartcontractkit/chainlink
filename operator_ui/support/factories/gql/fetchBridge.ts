// buildBridgePayloadFields builds the bridge fields.
export function buildBridgePayloadFields(
  overrides?: Partial<BridgePayload_Fields>,
): BridgePayload_Fields {
  return {
    __typename: 'Bridge',
    id: 'bridge-api',
    name: 'bridge-api',
    url: 'http://bridge.com',
    confirmations: 1,
    minimumContractPayment: '0',
    outgoingToken: 'outgoing1',
    ...overrides,
  }
}
