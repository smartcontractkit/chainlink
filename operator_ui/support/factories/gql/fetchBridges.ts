// buildBridge builds a Bridge for the FetchBridges query.
export function buildBridge(
  overrides?: Partial<BridgesPayload_ResultsFields>,
): BridgesPayload_ResultsFields {
  return {
    __typename: 'Bridge',
    id: 'bridge-api',
    name: 'bridge-api',
    url: 'http://bridge.com',
    confirmations: 1,
    minimumContractPayment: '0',
    ...overrides,
  }
}

// buildsBridges builds a list of bridges.
export function buildBridges(): ReadonlyArray<BridgesPayload_ResultsFields> {
  return [
    buildBridge({
      id: 'bridge-api1',
      name: 'bridge-api1',
      url: 'http://bridge1.com',
      confirmations: 1,
      minimumContractPayment: '100',
    }),
    buildBridge({
      id: 'bridge-api2',
      name: 'bridge-api2',
      url: 'http://bridge2.com',
      confirmations: 2,
      minimumContractPayment: '200',
    }),
  ]
}
