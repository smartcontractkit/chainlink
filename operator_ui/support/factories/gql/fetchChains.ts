import isoDate, { MINUTE_MS } from 'test-helpers/isoDate'

// buildChains builds a chain for the FetchChains query.
export function buildChain(
  overrides?: Partial<ChainsPayload_ResultsFields>,
): ChainsPayload_ResultsFields {
  const minuteAgo = isoDate(Date.now() - MINUTE_MS)

  return {
    __typename: 'Chain',
    id: '5',
    enabled: true,
    createdAt: minuteAgo,
    ...overrides,
  }
}

// buildsChains builds a list of chains.
export function buildChains(): ReadonlyArray<ChainsPayload_ResultsFields> {
  const minuteAgo = isoDate(Date.now() - MINUTE_MS)

  return [
    buildChain({
      id: '5',
      enabled: true,
      createdAt: minuteAgo,
    }),
    buildChain({
      id: '42',
      enabled: true,
      createdAt: minuteAgo,
    }),
  ]
}
