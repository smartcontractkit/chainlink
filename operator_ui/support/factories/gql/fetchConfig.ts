// buildConfigItem builds the config item fields.
export function buildConfigItem(
  overrides?: Partial<Config_ItemsFields>,
): Config_ItemsFields {
  return {
    __typename: 'ConfigItem',
    key: 'BAND',
    value: 'Major Lazer',
    ...overrides,
  }
}
