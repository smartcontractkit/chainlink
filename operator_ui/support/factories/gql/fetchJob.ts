import isoDate, { MINUTE_MS } from 'test-helpers/isoDate'

// buildJob builds the job payload fields for the FetchJob query.
export function buildJob(
  overrides?: Partial<JobPayload_Fields>,
): JobPayload_Fields {
  const minuteAgo = isoDate(Date.now() - MINUTE_MS)

  return {
    __typename: 'Job',
    id: '1',
    type: 'directrequest',
    schemaVersion: 1,
    name: 'direct request job',
    externalJobID: '00000000-0000-0000-0000-0000000000001',
    maxTaskDuration: '10s',
    gasLimit: 1000,
    spec: {
      __typename: 'DirectRequestSpec',
      contractAddress: '0x0000000000000000000000000000000000000000',
      evmChainID: '42',
      minIncomingConfirmations: 3,
      minIncomingConfirmationsEnv: false,
      minContractPaymentLinkJuels: '100000000000000',
      requesters: ['0x59bbE8CFC79c76857fE0eC27e67E4957370d72B5'],
    },
    runs: {
      results: [
        {
          __typename: 'JobRun',
          id: '1',
          allErrors: [],
          status: 'COMPLETED',
          createdAt: minuteAgo,
          finishedAt: minuteAgo,
        },
      ],
      metadata: { total: 0 },
    },
    observationSource:
      '    fetch    [type=http method=POST url="http://localhost:8001" requestData="{\\"hi\\": \\"hello\\"}"];\n    parse    [type=jsonparse path="data,result"];\n    multiply [type=multiply times=100];\n    fetch -> parse -> multiply;\n',
    createdAt: minuteAgo,
    errors: [],
    ...overrides,
  }
}

export function buildRun(
  overrides?: Partial<JobPayload_RunsFields>,
): JobPayload_RunsFields {
  const minuteAgo = isoDate(Date.now() - MINUTE_MS)

  return {
    __typename: 'JobRun',
    id: '1',
    allErrors: [],
    status: 'COMPLETED',
    createdAt: minuteAgo,
    finishedAt: minuteAgo,
    ...overrides,
  }
}

export function buildError(
  overrides?: Partial<JobPayload_ErrorsFields>,
): JobPayload_ErrorsFields {
  const minuteAgo = isoDate(Date.now() - MINUTE_MS)
  const twoMinutesAgo = isoDate(Date.now() - MINUTE_MS * 2)

  return {
    __typename: 'JobError',
    description: 'no contract at address',
    occurrences: 1,
    createdAt: twoMinutesAgo,
    updatedAt: minuteAgo,
    ...overrides,
  }
}
