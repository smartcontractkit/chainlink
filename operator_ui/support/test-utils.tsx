import React from 'react'
import { Provider as ReduxProvider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'
import { render, RenderOptions, RenderResult } from '@testing-library/react'

import { MuiThemeProvider } from '@material-ui/core/styles'

import createStore from 'src/createStore'
import { theme } from 'src/theme'

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MuiThemeProvider theme={theme}>
      <ReduxProvider store={createStore()}>{children}</ReduxProvider>
    </MuiThemeProvider>
  )
}

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
): RenderResult =>
  render(ui, {
    wrapper: AllTheProviders,
    ...options,
  })
interface RenderWithRouterProps {
  initialEntries?: string[]
}

// renderWithRouter behaves like 'render' except it wraps the provided ui in a
// Router.
//
// Use this when you need a router in your tests for page components and any
// component which uses react-router hooks.
const renderWithRouter = (
  ui: React.ReactElement,
  { initialEntries }: RenderWithRouterProps = { initialEntries: ['/'] },
  options?: Omit<RenderOptions, 'wrapper'>,
): RenderResult => {
  return {
    ...customRender(
      <MemoryRouter initialEntries={initialEntries}>{ui}</MemoryRouter>,
      options,
    ),
  }
}

export * from '@testing-library/react'
export { customRender as render, renderWithRouter }
