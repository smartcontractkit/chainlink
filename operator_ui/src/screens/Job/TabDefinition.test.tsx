import * as React from 'react'

import { Route } from 'react-router-dom'
import { renderWithRouter, screen } from 'support/test-utils'
import userEvent from '@testing-library/user-event'

import { buildJob } from 'support/factories/gql/fetchJob'
import { TabDefinition } from './TabDefinition'

const { getByRole, getByTestId, queryByText } = screen

describe('TabOverview', () => {
  function renderComponent(job: JobPayload_Fields) {
    renderWithRouter(
      <>
        <Route exact path="/jobs/:id/definition">
          <TabDefinition job={job} />
        </Route>
      </>,
      { initialEntries: ['/jobs/1/definition'] },
    )
  }

  it('renders the definition', () => {
    const job = buildJob()

    renderComponent(job)

    expect(getByTestId('definition').textContent).toMatchSnapshot()
  })

  it('can copy the definition', () => {
    // The copy package used window.prompt to copy to clipboard
    window.prompt = jest.fn()

    const job = buildJob()

    renderComponent(job)

    userEvent.click(getByRole('button', { name: /copy/i }))

    expect(queryByText(/copied!/i)).toBeInTheDocument()
    expect(window.prompt).toHaveBeenCalled()
  })
})
