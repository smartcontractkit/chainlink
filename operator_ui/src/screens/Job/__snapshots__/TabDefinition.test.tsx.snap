// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabOverview renders the definition 1`] = `
"type = \\"directrequest\\"
schemaVersion = 1
name = \\"direct request job\\"
externalJobID = \\"00000000-0000-0000-0000-0000000000001\\"
gasLimit = 1_000
maxTaskDuration = \\"10s\\"
contractAddress = \\"0x0000000000000000000000000000000000000000\\"
evmChainID = \\"42\\"
minIncomingConfirmations = 3
minContractPaymentLinkJuels = \\"100000000000000\\"
requesters = [ \\"0x59bbE8CFC79c76857fE0eC27e67E4957370d72B5\\" ]
observationSource = \\"\\"\\"
    fetch    [type=http method=POST url=\\"http://localhost:8001\\" requestData=\\"{\\\\\\\\\\"hi\\\\\\\\\\": \\\\\\\\\\"hello\\\\\\\\\\"}\\"];
    parse    [type=jsonparse path=\\"data,result\\"];
    multiply [type=multiply times=100];
    fetch -> parse -> multiply;
\\"\\"\\"

"
`;
